A modified Stillinger-Weber potential for transition metal dichalcogenide

oywg11 opened the pull request:
**Summary**

Adds a modification to the Stillinger-Weber potential that developed for transition metal dichalcogenides.

**Related Issue(s)**

<!--If this addresses an open GitHub issue for this project, please mention the issue number here, and describe the relation. Use the phrases `fixes #221` or `closes #135`, when you want an issue to be automatically closed when the pull request is merged-->

**Author(s)**

Jin-Wu Jiang, Shanghai University, jwjiang5918@hotmail.com 
Wengen Ouyang, Wuhan University, w.g.ouyang@gmail.com 

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

<!--Please state whether any changes in the pull request will break backward compatibility for inputs, and - if yes - explain what has been changed and why-->

**Implementation Notes**

This potential is designed for simulations of materials when distinguishing three-body angles are necessary, such as borophene and transition metal dichalcogenide (TMD), which cannot be described by the original code for the Stillinger-Weber potential. Validation, benchmark tests, and applications of this potential can be found in Nanotechnology 26, 315706 (2015) and Acta Mech. Solida. Sin 32, 17 (2019).

The Stillinger-Weber potential is widely used for materials with different angles around each atom, but the present version in LAMMPS cannot differentiate different angles in the original code for Stillinger-Weber potential. As a result, this modification has to be independently implemented by many LAMMPS users, who have contacted Prof. Jin-Wu Jiang for details about this modification (two most recent and two earlier emails are shown in the following figure). It can be much more convenient for LAMMPS users if this modification can be implemented in the distributed code of LAMMPS.

A flag is introduced to turn on/off the modification.
 

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

![image](https://user-images.githubusercontent.com/34092370/143885804-d3861f7a-57aa-49f0-829b-f99609f6919b.png)
http://stacks.iop.org/0957-4484/26/i=31/a=315706
https://doi.org/10.1007/s10338-018-0049-z

akohlmey commented:
@oywg11 thanks for your submission. However, this modification cannot be accepted in this form as it will silently give incorrect results for all accelerated versions of the "sw" pair style. My suggestion would be to follow the precedent of the "tersoff" potential, where there is a "tersoff/mod" pair style implemented as a derived class with modified versions of some of the functions used by the base PairTersoff class. The name of that new pair style could be something like "sw/angle" or "sw/mod".
Some more detailed comments are added to individual changes.

akohlmey commented on doc/src/pair_sw.rst:
```diff
@@ -14,7 +14,15 @@ Syntax
-   pair_style sw
+   pair_style style keywords values
+
+* style = *sw* or *sw/gpu* or *sw/omp*
```
We usually don't list individual accelerated pair styles like this. See, e.g. the `pair_airebo.rst` documentation file, which documents multiple variants of AIREBO *and* accelerated variants.

oywg11 replied:
Removed these changes.

akohlmey commented on doc/src/pair_sw.rst:
```diff
@@ -14,7 +14,15 @@ Syntax
+  .. parsed-literal::
+
+       *modify* value = flag
+         flag = 0/1 to turn off/on the modification
```
When this functionality is implemented as a derived class, no keyword would be needed.

oywg11 replied:
Removed. See the new codes.

akohlmey replied:
Mind you with derived classes, it would also be trivial to add OPENMP package support (because you can do the same threebody() function replacement as with the MANYBODY version, just derive from the OPENMP pair style).

oywg11 replied:
Thanks for the suggestion. I added the sw/mod/omp style and the corresponding examples.

akohlmey replied:
There were multiple problems with the sw/mod/omp style: it was added to the wrong folder (MANYBODY instead of OPENMP), the /omp version was derived from sw/mod, but instead it should have been derived from sw/omp, which results in a much simpler and smaller (and more effective) implementation.

akohlmey commented on doc/src/pair_sw.rst:
```diff
@@ -157,16 +168,52 @@ used for anything and can be set to 0.0 if desired.
 ----------
 
-.. include:: accel_styles.rst
+Styles with a *gpu*, *intel*, *kk*, *omp*, or *opt* suffix are
```
This change reverts changes in upstream. Please undo.

oywg11 replied:
Removed these changes and add a new file "pair_sw_mod.rst".

akohlmey replied:
Sorry, I don't see this properly restored. Perhaps you missed committing all changes or pushing all of them.
Besides, I would prefer to have just one .rst file for pair styles sw and sw/mod. Pair style tersoff/mod is not a good template to follow here. Overall, we are trying to balance manual pages to avoid having too many with lots of redundant content, but also split those with too many variants in the same file, so that we reach a balance between better maintainability (i.e. fewer files) and better readability (i.e. less convoluted pages)

oywg11 replied:
Thanks for the check, I indeed forgot to commit all changes. This issue is now fixed and the rst files are merged into a single one.

akohlmey replied:
There were a few more changes missing. I have updated the PR to properly include and link the docs, reordered the paragraphs and fixed a bunch more things.

akohlmey commented on examples/sw/in.mos2:
```diff
@@ -0,0 +1,33 @@
+processors	* * 1
+
+atom_style 	full
+read_data       ./single_layer_MoS2.data
```
Please remove the "./" for increased portability

oywg11 replied:
Fixed.

akohlmey commented on examples/sw/in.mos2:
```diff
@@ -0,0 +1,33 @@
+variable        mS  equal 32.065   # mass of sulphur atom , uint: a.u.=1.66X10^(-27)kg
+variable        mMo equal 95.94    # mass of molebdenum atom , uint: a.u.=1.66X10^(-27)kg
+
+mass            * ${mS}
```
Please don't use variables for such trivial cases. That just makes the input harder to read.

oywg11 replied:
Fixed.

akohlmey commented on examples/sw/tmd.sw:
```diff
@@ -0,0 +1,143 @@
+# DATE: 2018-03-26 UNITS: metal CONTRIBUTOR: Jin-Wu Jiang jwjiang5918@hotmail.com
```
Because of the requested renaming of the pair style, this potential file should also have a different suffix

oywg11 replied:
renamed tmd.sw to tmd.sw.mod.

akohlmey replied:
renamed but not deleted. this is fixed now, too.

akohlmey commented on examples/sw/log.27Oct21.mos2.icc.1.lmp:
```diff
@@ -0,0 +1,109 @@
+LAMMPS (27 Oct 2021)
```
Please see the file name conventions for log files as they are used in other examples.

oywg11 replied:
Renamed the files.

akohlmey commented on examples/sw/tmd.sw:
```diff
@@ -0,0 +1,143 @@
+# DATE: 2018-03-26 UNITS: metal CONTRIBUTOR: Jin-Wu Jiang jwjiang5918@hotmail.com
+# CITATION: J.-W. Jiang, Acta Mech. Solida. Sin 32, 17 (2019).
```
Potential files in example folders should be symbolic links to their copy in the potentials folder to make certain they stay consistent.

oywg11 replied:
> @oywg11 thanks for your submission. However, this modification cannot be accepted in this form as it will silently give incorrect results for all accelerated versions of the "sw" pair style. My suggestion would be to follow the precedent of the "tersoff" potential, where there is a "tersoff/mod" pair style implemented as a derived class with modified versions of some of the functions used by the base PairTersoff class. The name of that new pair style could be something like "sw/angle" or "sw/mod". Some more detailed comments are added to individual changes.

@akohlmey Following your suggestion, I modified this potential as a derived class of the original SW potential. However, I met an issue that needs your help. Since the modification is only in the three-body term, I kept it only in the derived class following the same structure as pair_tersoff_mod. The issue is that when I run simulations with sw/mod, I found it still uses the original three-body term instead of the modified one. Thus there must be something important that I didn't implement but I cannot find it. Could you give me some suggestions? Thanks!

oywg11 replied:
> 

I think I solved this issue.

oywg11 replied:
> 

I worked on this pull request in a virtual Linux machine and I failed to generate a link to the potential file. Could you fix this issue? Thanks!

akohlmey commented on src/MANYBODY/pair_sw.cpp:
```diff
@@ -231,9 +233,24 @@ void PairSW::allocate()
+void PairSW::settings(int narg, char **arg)
 {
-  if (narg != 0) error->all(FLERR,"Illegal pair_style command");
+  //if (narg != 0) error->all(FLERR,"Illegal pair_style command");
```
This block of code should not be needed for a derived class.

oywg11 commented:
@akohlmey Thanks for the suggestion! I will modify the files accordingly.

akohlmey commented on src/MANYBODY/pair_sw.cpp:
```diff
@@ -292,8 +292,22 @@ void PairSW::read_file(char *file)
   if (comm->me == 0) {
-    PotentialFileReader reader(lmp, file, "sw", unit_convert_flag);
-    char *line;
+    std::string potential_name;
```
In this case I think the "variant" flag would be overkill unless there is also a requirement to parse the potential files differently, and/or a check whether the potential file would be compatible with a specific SW variant.
From what I read in this PR, the potential files seem to be compatible with either pair style variant, just the actual potential function that is applied is slightly different.

oywg11 replied:
The "variant" feature is removed.

akohlmey replied:
thanks,

sjplimp commented:
Fine with me but please wait for @athomps to review

athomps reviewed (changes requested):
Thank you for the submission. I have one serious concern about this submission and 5 minor ones.

The serious concern is whether the forces are consistent with the energy expression.  The later is changed in pair_sw_mod.cpp, but the former appears to use the original pair_sw.cpp force expressions.  Given that the new energy expression contains a new Cosine function, it is unlikely that the forces are still correct.

Here are the minor concerns:

1. The first citation (2015) makes no mention of a modified SW angle term, instead reports results only for standard SW.
2. The second citation (2019) is behind a paywall that I do not have access to
3. The equation provided in the documentation contains undefined variables: g_C, delta_1, delta_2
4. The relevant block of code does not match the expression in the documentation e.g. use of abs(delta) is not documented, and also variables replaced with hardcoded values
5. Hard-coded values should at a minimum be replaced with named constants assigned at compile time

athomps commented:
I examined the example and now understand what the real intent of this modification is.  The distribution of bond angles in MoS2 is bimodal, with one peak at costheta = 0.143 and another at costheta ~ -0.7.  To handle this, it is desired to ignore any bond angles with costheta < -0.5.  The submitted code achieves this, but it is not done in a very clear or general way.  Also, because the switching function on the hardcoded range of costheta is not reflected in the forces, the switching function has no effect on dynamics and serves no useful purpose.

athomps commented:
A better way to accomplish the same result would be to add a keyword for pair_style sw e.g. `maxdelcostheta 0.2` that will set threebody energy to zero for all bond angles whose cosine deviates by more than 0.2 from costheta0.  This would require adding a single 0/1 if-test to `PairSW::threebody()` which I think should not be too much of a slowdown for base case.

```
  delcs = cs - paramijk->costheta;
  if(maxdelcosthetaflag && fabs(delcs) >= maxdelcos) {
       fj[0] = fj[1] = fj[2] = fk[0] = fk[1] = fk[2] = eng = 0.0;
       return;
  }

```

akohlmey commented:
> add a keyword for pair_style sw

It would still be better to have that in the derived sw/mod pair style since the feature does not exist in many of the accelerated versions and thus would result in inconsistent behavior. This is why I pushed for making this modification a derived class instead of a keyword.

athomps commented:
I see.  What about preventing the silent incorrect results by adding an error check at initialization "ERROR: keyword maxdelcostheta not supported by pair_style sw/kk" ?

akohlmey commented:
> I see.  What about preventing the silent incorrect results by adding an error check at initialization "ERROR: keyword maxdelcostheta not supported by pair_style sw/kk" ?
> 

Would be messier and more programming. We've something similar with the shift keyword with tersoff and it is not nice. It was easier to accept because there are so many different tersoff variants that share code. Sw is currently much simpler.

athomps commented:
Ok. Then I am good with sw/mod, but it should be done in a way that is more direct and much less cryptic than the current submission e.g. using a keyword and a user-specified threshold on `delcss`. Also, the threshold should be defined by a simple equation in the documentation that uses the same notation as existing threebody equation. The 2015 paper seems to be based on the modified LAMMPS code, but that is not stated in the paper, which is a little frustrating.

jwjiang5918 commented:
Hi all ! This is Jin-Wu Jiang, the author of 2015 and 2019 papers. Thanks for your suggestions.

About five years ago, I suggested to add this modification to LAMMPS. It was not accepted at that time, since only few researchers started to use SW in LAMMPS for structures like MoS2. However, more and more users contacted me about resolving this issue after my 2015 paper, so Wen-Gen and I decided to request this modification.

My 2015 paper uses a modified LAMMPS, with a threshhold value for the deltacs. This is noted at the end of the paper, just before the conclusion section.

My 2019 paper modifies LAMMPS by using a continuous cut-off function for deltacs (see figures for the expression and the curve for the cut-off function). I suggested this modification (with hard coded values), because it does not introduce any new quantity to pair-sw.cpp. Users can realize their purpose with minimum modification for LAMMPS.

@athomps For the serious concern of athomps "The serious concern is whether the forces are consistent with the energy expression", the purpose of the cut-off function is to turn off the SW for both potential and force if delcs is too large, both of which are realized in the pair-sw.cpp file (as delcs==0 if delcs>=0.35). Hence, both SW potential and SW force will be turned off if delcs>=0.35.

Agree that, the threshhold value should be given by users.

![cut-off-expression](https://user-images.githubusercontent.com/22739845/144614397-427a9551-8a50-49b4-9031-ca1ba2ea36c3.png)
![cut-off-expression](https://user-images.githubusercontent.com/22739845/144614543-eeefd1bc-92a5-49c8-b1b6-7005afdab5a6.png)

jwjiang5918 commented:
![cut-off-curve](https://user-images.githubusercontent.com/22739845/144615125-9af6b4a1-da25-4d8a-9f13-40b40111c6ed.png)

oywg11 commented:
> Ok. Then I am good with sw/mod, but it should be done in a way that is more direct and much less cryptic than the current submission e.g. using a keyword and a user-specified threshold on `delcss`. Also, the threshold should be defined by a simple equation in the documentation that uses the same notation as existing threebody equation. The 2015 paper seems to be based on the modified LAMMPS code, but that is not stated in the paper, which is a little frustrating.

I tried to implement the changes you requested, for the normal sw/mod style, it works fine. But for sw/mod/omp style, it seems that more programming is needed. At the present stage, I implement the user-specified threshold for the former and named constants for the latter.

akohlmey commented:
@oywg11 I've updated the code to compile cleanly and to be consistent between MANYBODY and OPENMP version. There was a call to the settings() function of the base class missing (it doesn't do much, but without it all additional arguments will be ignored instead of causing an error). I think this may still need some more detail in the documentation to mention the possible values and requirements (delta1 must be smaller than delta2 obviously). Plus there are the concerns from @athomps that should be discussed.

athomps commented:
Looks good to me.  Here are some optional changes:

1. My example files from examples/sw to examples/threebody
2. Change atom style to atomic and get rid of molecule and charge columns from data file
3. The equations for delta and f_c are garbled. I suggest just providing an equation for f_c, preceded by "The value of $\delta$ used in the original energy and force expression is scaled by a switching factor $f_c$ that decreases smoothly from 1 to 0 over the range [delta1,delta2].  This smoothly turns off the energy and force contributions for $|\delta|$ greater than delta 2, although this is not done in a way that fully conserves energy."

oywg11 commented:
> @oywg11 I've updated the code to compile cleanly and to be consistent between MANYBODY and OPENMP version. There was a call to the settings() function of the base class missing (it doesn't do much, but without it all additional arguments will be ignored instead of causing an error). I think this may still need some more detail in the documentation to mention the possible values and requirements (delta1 must be smaller than delta2 obviously). Plus there are the concerns from @athomps that should be discussed.

@akohlmey Many thanks for the correction! I will discuss with @jwjiang5918 for a more detailed documentation.

oywg11 commented:
> Looks good to me. Here are some optional changes:
> 
> 1. My example files from examples/sw to examples/threebody
> 2. Change atom style to atomic and get rid of molecule and charge columns from data file
> 3. The equations for delta and f_c are garbled. I suggest just providing an equation for f_c, preceded by "The value of $\delta$ used in the original energy and force expression is scaled by a switching factor $f_c$ that decreases smoothly from 1 to 0 over the range [delta1,delta2].  This smoothly turns off the energy and force contributions for $|\delta|$ greater than delta 2, although this is not done in a way that fully conserves energy."

@athomps Thanks for the further suggestions. I will make the corresponding changes.

oywg11 commented:
@akohlmey @athomps We updated the documentation and example files. Please let us know if you have further comments or suggestions. Thanks!

athomps commented:
This all looks good to me. There were sphinx :math: errors in the new inline expressions. I fixed them, but could not push the changes, should be:

`
This cut-off function decreases smoothly from 1 to 0 over the range :math:`[\delta_1, \delta_2]`.
This smoothly turns off the energy and force contributions for :math:`\lvert \delta \rvert > \delta_2`.
It is suggested that :math:`\delta_1` and :math:`\delta_2` to be the value around
:math:`0.5 \lvert \cos \theta_1 - \cos \theta_2 \rvert`,
with :math:`\theta_1` and :math:`\theta_2` as the different types of angles around an atom.
For borophene and transition metal dichalcogenide, :math:`\delta_1 = 0.25` and :math:`\delta_2 = 0.35`.
`

Also, we should still get rid of examples/sw and move example to examples/threebody

oywg11 commented:
> This all looks good to me. There were sphinx :math: errors in the new inline expressions. I fixed them, but could not push the changes, should be:
> 
> ` This cut-off function decreases smoothly from 1 to 0 over the range :math:`[\delta_1, \delta_2]`. This smoothly turns off the energy and force contributions for :math:`\lvert \delta \rvert > \delta_2`. It is suggested that :math:`\delta_1` and :math:`\delta_2` to be the value around :math:`0.5 \lvert \cos \theta_1 - \cos \theta_2 \rvert`, with :math:`\theta_1` and :math:`\theta_2` as the different types of angles around an atom. For borophene and transition metal dichalcogenide, :math:`\delta_1 = 0.25` and :math:`\delta_2 = 0.35`. `
> 
> Also, we should still get rid of examples/sw and move example to examples/threebody

@athomps Thanks! I fixed these issues in the documentation and moved the example files to examples/threebody.

akohlmey commented:
@oywg11 @athomps I've corrected multiple formatting/syntax/spelling issues with the documentation changes. This now passes all formal tests and requirements for merging.

oywg11 commented:
> @oywg11 @athomps I've corrected multiple formatting/syntax/spelling issues with the documentation changes. This now passes all formal tests and requirements for merging.

@akohlmey Thanks for the corrections! The new version looks good to me.

athomps reviewed (approved):
I approve
